File: mercado.js
class Produto {
    constructor(nome, valor, validade) {
        this.nome = nome;
        this.valor = valor;
        this.validade = new Date(validade);
    }

    verificarValidade(dataAtual) {
        const dataAtualDate = new Date(dataAtual);
        if (this.validade < dataAtualDate) {
            const diferenca = Math.ceil((dataAtualDate - this.validade) / (1000 * 60 * 60 * 24));
            return `Produto ${this.nome} venceu há ${diferenca} dias.`;
        }
        return `Produto ${this.nome} está dentro da validade.`;
    }
}
// Lista de produtos
const listaProdutos = [
    new Produto("Arroz", 20.00, "2026-12-31"),
    new Produto("Feijão", 10.00, "2026-11-30"),
    new Produto("Macarrão", 5.00, "2026-10-15"),
    new Produto("Batata", 3.25, "2026-09-10"),
    new Produto("Tomate", 4.50, "2026-09-20"),
    new Produto("Cebola", 2.75, "2026-10-05"),
    new Produto("Alface", 1.50, "2026-09-25"),
    new Produto("Maçã", 3.00, "2026-10-01"),
    new Produto("Banana", 2.00, "2026-09-28"),
    new Produto("Leite", 4.00, "2026-09-30")
];
// Função principal
function caixaSupermercado(dataAtual, saldo, itensComprados) {
    let total = 0;
    let produtosComprados = [];

    itensComprados.forEach(item => {
        const produto = listaProdutos[item];
        console.log(produto.verificarValidade(dataAtual));
        total += produto.valor;
        produtosComprados.push(produto.nome);
    });

    console.log(`Total da compra: R$${total.toFixed(2)}`);

    if (total > saldo) {
        console.log(`Saldo insuficiente. Faltam R$${(total - saldo).toFixed(2)}.`);
    } else {
        console.log(`Compra realizada com sucesso! Troco: R$${(saldo - total).toFixed(2)}.`);
        console.log(`Produtos comprados: ${produtosComprados.join(", ")}`);
    }
}
// Exemplo de uso
let dataAtual = "2024-09-24";
let saldo = 50.00;
let itensComprados = [0, 1, 2]; 
// Índices dos produtos na lista
caixaSupermercado(dataAtual, saldo, itensComprados);

File: mercado.dart
import 'dart:io';
class Produto {
  String? nome;
  double? valor;
  DateTime? validade;

  Produto(this.nome, this.valor, this.validade);
}
//listando os produtoso com (Nome, preço, Data de Validade)
//A data de validade deverá ser por formato ISO (yyyy/mm/dd)
void main() {
  List<Produto> listaDeProdutos = [
    Produto('Arroz', 20.0, DateTime(2024, 10, 1)),
    Produto('Feijão', 7.5, DateTime(2024, 9, 15)),
    Produto('Macarrão', 5.0, DateTime(2024, 11, 10)),
    Produto('Leite', 4.5, DateTime(2024, 9, 25)),
    Produto('Pão', 3.0, DateTime(2024, 9, 20)),
    Produto('Azeite', 6.0, DateTime(2024, 10, 5)),
    Produto('Queijo', 15.0, DateTime(2024, 10, 15)),
    Produto('Mortadela', 12.0, DateTime(2024, 9, 30)),
    Produto('Café', 8.0, DateTime(2024, 12, 1)),
    Produto('Sal', 4.0, DateTime(2025, 1, 1)),
  ];

  print('Digite a data de hoje (yyyy-mm-dd):');
  String? dataHojeStr = stdin.readLineSync();
  DateTime dataHoje = DateTime.parse(dataHojeStr!);

  print('Digite seu saldo:');
  double saldo = double.parse(stdin.readLineSync()!);
//Quantidade de produtos a ser realizado o pedido
//Necessário que seja um valor númerico decimal
  print('Quantos itens você quer comprar ?');

  int quantidadeItens = int.parse(stdin.readLineSync()!);

  double totalCompra = 0.0;
  List<Produto> itensComprados = [];
// Lista com o nome dos produtos disponiveis
print('Nós Temos:\nArroz \nFeijão \nMacarrão \nLeite \nPão \nAzeite \nQueijo \nMortadela \nCafé \nSal');
  for (int i = 0; i < quantidadeItens; i++) {
    print('Digite o nome do produto:');
    String? nomeProduto = stdin.readLineSync();
    Produto? produto = listaDeProdutos.firstWhere(
        (p) => p.nome!.toLowerCase() == nomeProduto!.toLowerCase(),
        orElse: () => Produto('', 0.0, DateTime.now()));

    if (produto.nome!.isEmpty) {
      print('Produto não encontrado.');
      i--;
      continue;
    }

    if (produto.validade!.isBefore(dataHoje)) {
      Duration diferenca = dataHoje!.difference(produto.validade!);
      print(
          'O produto ${produto.nome} está vencido há ${diferenca.inDays} dias.');
      i--;
      continue;
    }

    totalCompra += produto.valor!;
    itensComprados.add(produto);
  }
  if (totalCompra > saldo) {
    print('Saldo insuficiente. Faltam R\$${(totalCompra - saldo).toStringAsFixed(2)}.');
  } else {
    double troco = saldo - totalCompra;
    print('Compra realizada com sucesso!');
    print('Itens comprados:');
    for (var item in itensComprados) {
      print('${item.nome} - R\$${item.valor!.toStringAsFixed(2)}');
    }
    print('Troco: R\$${troco.toStringAsFixed(2)}');
  }
}
